Mingming/pg debezium cdc (#1426)
aiven/devportal: docs/products/postgresql/howto/postgresql-debezium-cdc.rst

> Handling PostgreSQL® node replacements when using Debezium for Change Data Capture

Reviewer: Please check the warning, we should have either titles in sentence case or, if there's an exception, we need to add the exception (like Debezium or Change Data Capture) to `.github/vale/styles/Aiven/capitalization_headings.yml`

Reviewer: also the sentence should be in present case, i would suggest

```
Handle PostgreSQL® node replacements when using Debezium for Change Data Capture
```

=================================================================================

> As an Aiven customer you may be running a Aiven for Apache Kafka® Connect

Reviewer: We can reference the devportal article about debezium doc https://docs.aiven.io/docs/products/kafka/kafka-connect/howto/debezium-source-connector-pg.html

cluster that runs one or more `Debezium PostgreSQL® source
connectors <https://debezium.io/documentation/reference/connectors/postgresql.html>`__
(Debezium) capturing changes from an Aiven for PostgreSQL® (PG) service.

After the PG service has had a maintenance update or undergo any
operation which replaces the nodes (such as a plan or cloud region
change), Debezium would lose connection to the PG database. It is able
to recover from connection errors to the database and start listening
for change events again after a restart of the connector's task.
However, in some cases the Debezium replication slot is lagging behind
as the amount of WAL logs grows causing a concern.

> There is a GitHub repository where you can easily spin up a Debezium -

Reviewer: Can this be a tip?

PostgreSQL setup to test the node replacement scenario and also follow
along this help article:
https://github.com/aiven/debezium-pg-kafka-connect-test

The possible errors encountered, the cause and solution to the growing
replication slot lag, and how to ensure a graceful PG node replacement
or maintenance upgrade are explained in the next sections.

> .. _h_ab9933dea8:

Reviewer: this anchor should be removed

> Possible errors seen in Debezium connector logs when PG has a node replacement

Reviewer: ```suggestion
Debezium errors related to PostgreSQL node replacement
```

------------------------------------------------------------------------------

> During the reported PG node replacements where Debezium could not

Reviewer: let's try to reference PostgreSQL always consistently and entirely, without abbreviations

recover, we have observed the following errors that have occurred in one
or more occasions:

-  ``org.apache.kafka.connect.errors.ConnectException: Could not create PG connection``

-  ``io.debezium.DebeziumException: Could not execute heartbeat action (Error: 57P01)``

-  ``org.PostgreSQL.util.PSQLException: ERROR: replication slot "SLOT_NAME" is active for PID xxxx``

These are all unrecoverable errors, meaning that they require a restart

> of the connector task(s) in order for the connector to resume operations

Reviewer: ```suggestion
of the connector task(s) in order to resume operations
```

> again. A restart can be performed manually either through the web

Reviewer: ```suggestion
again. A restart can be performed manually either through the `Aiven Console <https://console.aiven.io/>`_, in under the `Connectors` tab
```

console or via the `Kafka Connect REST
API <https://docs.confluent.io/platform/current/connect/references/restapi.html#rest-api-task-restart>`__

> (you can get the service URI from the web console).

Reviewer: ```suggestion
. You can get the service URI from the `Aiven Console <https://console.aiven.io/>`_, in the service detail page.
```

> .. image:: /images/products/postgresql/pg-debezium-cdc_image1.png

Reviewer: I'm not sure if this image is what we need or a better suggestion in the explanation

> For automatically restarting tasks due to unrecoverable errors, we can

Reviewer: this should be a `Tip`

> set *``"_aiven.restart.on.failure": true``* in the connector's

Reviewer: ```suggestion
set ``"_aiven.restart.on.failure": true`` in the connector's
```

configuration ( `help
article <https://help.aiven.io/en/articles/5088396-kafka-connect-auto-restart-on-failures>`__
). By default tasks are checked for errors every 15 minutes but we can
set this to something else. Please contact Aiven support if you would
like to have this check interval shortened.

> .. _h_09f7114538:

Reviewer: remove not needed anchor

> Debezium connectors/tasks have been restarted, but there is growing replication lag

Reviewer: ```suggestion
Growing replication lag after Debezium connector restart
```

-----------------------------------------------------------------------------------

> Per the Debezium docs, there are two reasons why this can happen (copied

Reviewer: I would replace `this` by what it can happen

> verbatim from the `Debezium

Reviewer: I wouldn't put the verbatim but rather trying to rephrase minimally and reference the doc

docs <https://debezium.io/documentation/reference/1.5/connectors/postgresql.html#postgresql-wal-disk-space>`__
):

#. *There are many updates in a database that is being tracked but only
   a tiny number of updates are related to the table(s) and schema(s)
   for which the connector is capturing changes. This situation can be
   easily solved with periodic heartbeat events. Set the
   ``heartbeat.interval.ms`` connector configuration property.*

#. *The PostgreSQL instance contains multiple databases and one of them
   is a high-traffic database. Debezium captures changes in another
   database that is low-traffic in comparison to the other database.
   Debezium then cannot confirm the LSN
   (*\ ```confirmed_flush_lsn`` <https://www.postgresql.org/docs/13/view-pg-replication-slots.html>`__\ *)
   as replication slots work per-database and Debezium is not invoked.
   As WAL is shared by all databases, the amount used tends to grow
   until an event is emitted by the database for which Debezium is
   capturing changes.*

> During basic testing, this has been observed to happen in 2 scenarios:

Reviewer: What is a basic testing?

Reviewer: can we either be more explicit or just call it testing?

#. The table(s) which the Debezium connector is tracking has not had any
   changes, and heartbeats are not enabled in the connector.

#. The table(s) which the Debezium connector is tracking has not had any
   changes, heartbeats are enabled (via
   ```heartbeat.interval.ms`` <https://debezium.io/documentation/reference/1.5/connectors/postgresql.html#postgresql-property-heartbeat-interval-ms>`__
   and
   ```heartbeat.action.query`` <https://debezium.io/documentation/reference/1.5/connectors/postgresql.html#postgresql-property-heartbeat-action-query>`__
   ), but the connector is not sending the heartbeat. Debezium not
   sending the heartbeat is a known bug reported in
   `DBZ-3746 <https://issues.redhat.com/browse/DBZ-3746>`__ .

.. _h_7415120456:

> Clearing the replication lag

Reviewer: ```suggestion
Clear the replication lag
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To clear the lag, simply resume the database activities (assuming you
have paused all traffic to the database) or make some changes to any
table in the database in which Debezium is capturing changes to invoke
the replication slot (such as the heartbeat table). Debezium would then
confirm the latest LSN and allow the database to reclaim the WAL space.

.. _h_b915a23266:

> Ensuring Debezium gracefully survives a PostgreSQL node replacement

Reviewer: ```suggestion
Ensure Debezium gracefully survives a PostgreSQL node replacement
```

-----------------------------------------------------------------

To prevent data loss (Debezium missing change events) in the event of a
PostgreSQL node replacement and ensure the automatic recovery of the
connector, here are several guidelines which you can follow.

.. _h_a711a06482:

> Immediate solutions

Reviewer: What is this immediate solution and why should one use it? I believe the title needs a bit more context

------------------------------------------------------------------------------

During the reported PostgreSQL node replacements where Debezium could not recover, we have observed the following errors that have occurred in one or more occasions::

   # ERROR 1
   org.apache.kafka.connect.errors.ConnectException: Could not create PostgreSQL connection
   # ERROR 2
   io.debezium.DebeziumException: Could not execute heartbeat action (Error: 57P01)
   # ERROR 3
   org.PostgreSQL.util.PSQLException: ERROR: replication slot "SLOT_NAME" is active for PID xxxx

These are all unrecoverable errors, meaning that they require a restart of the connector task(s) in order to resume operations again. A restart can be performed manually either through the `Aiven Console <https://console.aiven.io/>`_, in under the `Connectors` tab console or via the `Kafka Connect REST API <https://docs.confluent.io/platform/current/connect/references/restapi.html#rest-api-task-restart>`__. You can get the service URI from the `Aiven Console <https://console.aiven.io/>`_, in the service detail page.

.. image:: /images/products/postgresql/pg-debezium-cdc_image1.png

>    :alt: The Aiven Console page showing the Debezium connector error

Reviewer: @zhangmmaiven I believe the error was related to the missing `:alt:` for the image
